internal/dcache/file_manager/file_manager.go

```diff
@@ -0,0 +1,372 @@
+   copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+   SOFTWARE
+*/
+
+package filemanager
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"sync"
+	"syscall"
+
+	"github.com/Azure/azure-storage-fuse/v2/common"
+	"github.com/Azure/azure-storage-fuse/v2/common/log"
+	"github.com/Azure/azure-storage-fuse/v2/internal/dcache/file_manager/models"
+)
+
+const (
+	cacheAccessAzure uint16 = iota
+	cacheAccessDcache
+)
+
+type fileIOManager struct {
+	numReadAheadChunks int // Number of chunks to readahead from the current chunk.
+	numStagingChunks   int // Max Number of chunks per file that can be staging area at any time.
+	wp                 *workerPool
+	bp                 *bufferPool
+}
+
+var fileIOMgr fileIOManager
+
+func NewFileIOManager(workers int, numReadAheadChunks int, numStagingChunks int, bufSize int, maxBuffers int) {
+	fileIOMgr = fileIOManager{
+		numReadAheadChunks: numReadAheadChunks,
+		numStagingChunks:   numStagingChunks,
+	}
+	wp := NewWorkerPool(workers)
+	wp.startWorkerPool()
+	bp := NewBufferPool(bufSize, maxBuffers)
+	fileIOMgr.wp = wp
+	fileIOMgr.bp = bp
+}
+
+func EndFileIOManager() {
+	fileIOMgr.wp.endWorkerPool()
+}
+
+type File struct {
+	FileMetadata    *models.FileMetadata
+	Access          common.BitMap16 // Represents the cache access type like azure, dcache, both.
+	lastWriteOffset int64           // Every new offset that come to write should be greater than lastWriteOffset
+	// Offset should be monotonically increasing while writing to the file.
+	Chunks sync.Map // Chunk Idx -> *chunk
+	// The above chunks are the outstanding chunks for the file.
+	// Those chunks can be readahead chunks for read /
+	// current staging chunks for write
+	// Todo: Chunks should be tracked globally rather than per file.
+}
+
+// Reads the file from the corresponding chunk(s) to the buf.
+func (file *File) ReadFile(offset int64, buf []byte) (bytesRead int, err error) {
+	log.Debug("DistributedCache::ReadFile : offset : %d, bufSize : %d, file : %s", offset, len(buf), file.FileMetadata.Filename)
+	if offset >= file.FileMetadata.Size {
+		return 0, io.EOF
+	}
+	endOffset := min(offset+int64(len(buf)), file.FileMetadata.Size)
+	bufOffset := 0
+	for offset < endOffset {
+		//  Currently calling direct readaahead for the chunk assuming sequential workflow
+		chnkIdx := getChunkIdxFromFileOffset(offset, file.FileMetadata.FileLayout)
+		// todo : Detection of the sequential nature of the handle can be found from storing the offset
+		// in the handle.
```

Comment:
Good seq/random pattern detection is more complicated than just comparing with the last offset read.
We can do it properly later. For now add a TODO asking for adding sequential pattern detection